### PR TITLE
[hailctl] update to dataproc 2.2 and Spark 3.5.0

### LINKED ISSUE #13971: [query] Support Spark 3.5.x
### What happened?

At time of writing, building hail with `SPARK_VERSION=3.4.0` errors with the following message:

```
hail/src/main/scala/is/hail/HailContext.scala:119:21: value implOpMulMatrix_DMD_DVD_eq_DVD is not a member of object breeze.linalg.DenseMatrix
```

This is due to a major version upgrade and breaking change in the Breeze library on which spark and hail depend. The exact error is a rename and refactor. The method `DenseMatrix.implOpMulMatrix_DMD_DVD_eq_DVD` is now `HasOps.impl_OpMulMatrix_DMD_DVD_eq_DVD`. Notice the method name change and the fact that `HasOps` does not exist in the version of Breeze (1.x) that is used in Spark 3.3.

Hail should build with Spark 3.4, but since we only officially support one version of Spark (whichever Dataproc currently is running), it would be reasonable to wait to fully upgrade to Spark 3.4 when [Dataproc 2.2](https://cloud.google.com/dataproc/docs/concepts/versioning/dataproc-release-2.2) is GA instead of trying to do something hacky to support both versions of Breeze.

### Version

0.2.126

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by danking**
Fixes #13971

CHANGELOG: Hail now supports and primarily tests against Dataproc 2.2.5, Spark 3.5.0, and Java 11. We strongly recommend updating to Spark 3.5.0 and Java 11. You should also update your GCS connector *after installing Hail*: `curl https://broad.io/install-gcs-connector | python3`. Do not try to update before installing Hail 0.2.131.

https://cloud.google.com/dataproc/docs/concepts/versioning/dataproc-release-2.2

### CODE REVIEW

File: hail/src/main/scala/is/hail/HailContext.scala
Comment (danking) [inline hail/src/main/scala/is/hail/HailContext.scala:129]: see https://github.com/hail-is/hail/issues/13971#issuecomment-1792870445

**Comment by danking** [comment]
@zyd14 that's just a warning message printed during compilation. The requirements.txt file [accepts any 3.5.x version](https://github.com/hail-is/hail/pull/14158/files#diff-c4e24762fa3ba8e53670cc9aba74121ade8ac055942738bde9c951fc0514561aR14).

**Comment by danking** [comment]
I had to pin nbconvert<7.14 due to a new bug in nbconvert.

https://github.com/jupyter/nbconvert/issues/2092

File: hail/python/hailtop/batch/batch_pool_executor.py
Comment (danking) [inline hail/python/hailtop/batch/batch_pool_executor.py:235]: the lint didn't like that I used an empty generator where an async generator was expected.

**Comment by danking** [comment]
Since the dataproc tests only run on main commits (not on every PR commit, due to cost), I submitted a dev deploy to test the latest commit to this branch against dataproc: https://ci.hail.is/batches/8119055

```
hailctl dev deploy -b danking/hail:dataproc-2.2 -s test_dataproc-37 -s test_dataproc-38
```

**Comment by danking** [comment]
Previous dev deploy passed. Rebased on main and submitted a new dev deploy: https://ci.hail.is/batches/8120683

**Review by ehigham — DISMISSED** [review]

File: hail/src/main/scala/is/hail/HailContext.scala
Comment (danking) [inline hail/src/main/scala/is/hail/HailContext.scala:87]: I don't fully grok what we were trying to accomplish here.

You can't possibly get to these lines if the JVM version is lower than the byte code, so JVMs <=7 would never execute this. AFAICT, Spark now supports all LTS versions of Java (currently: 8, 11, 17), so I see no strong reason to prohibit any version. Instead, I simply warn if you are using a version of the JVM other than the one with which we test.

**Comment by danking** [comment]
@ehigham ready for re-review now with Java 11 changes.

File: hail/src/main/scala/is/hail/HailContext.scala
Comment (danking) [reply hail/src/main/scala/is/hail/HailContext.scala:87]: In particular:

> Spark runs on Java 8/11/17, Scala 2.12/2.13, Python 3.8+, and R 3.5+. Java 8 prior to version 8u371 support is deprecated as of Spark 3.5.0.

https://spark.apache.org/docs/latest/#downloading

**Comment by danking** [comment]
And here's a dev deploy that runs the dataproc tests. Don't approve until these tests pass! We don't run them on ordinary PRs because they're expensive and slow. We do run them on main commits. For this PR, the chance of having broken dataproc is high enough that we should ensure the tests pass before merging into main.

https://ci.hail.is/batches/8121061

**Comment by danking** [comment]
Delay merging until https://github.com/broadinstitute/install-gcs-connector/pull/6 is merged. Without that PR, users will not have access to a version of the GCS Hadoop connector that does not use tons of memory in JVM 11.

**Comment by danking** [comment]
Rebased and dev deploy kicked off: https://ci.hail.is/batches/8122588

**Comment by danking** [comment]
https://github.com/broadinstitute/install-gcs-connector/pull/6 is merged.

File: hail/build.sc
Comment (patrick-schultz) [inline hail/build.sc:83]: It appears we've actually been using breeze 1.2, because spark-mllib pulls it in:
```
❯ mill ivyDepsTree --withCompile --withRuntime --whatDependsOn org.scalanlp:breeze_2.12                                                        (base)
[17/17] ivyDepsTree
└─ org.scalanlp:breeze_2.12:1.2
   ├─ org.apache.spark:spark-mllib-local_2.12:3.3.0
   │  ├─ org.apache.spark:spark-graphx_2.12:3.3.0
   │  │  └─ org.apache.spark:spark-mllib_2.12:3.3.0
   │  └─ org.apache.spark:spark-mllib_2.12:3.3.0
   ├─ org.apache.spark:spark-mllib_2.12:3.3.0
   └─ org.scalanlp:breeze-natives_2.12:1.1 org.scalanlp:breeze_2.12:1.1 -> 1.2
```

Comment (danking) [reply hail/build.sc:83]: Does the compileIvyDeps line not exclude them? I want to let Spark use their thing and relocate breeze for our own internal purposes.

Comment (patrick-schultz) [reply hail/build.sc:83]: Ah, I missed that, that does let us use breeze 1.1. And I now realize even without that, because this is a compile only dependency, we would build with breeze 1.2 but run with 1.1.

**Comment by danking** [comment]
@patrick-schultz @ehigham I'm abdicating responsibility for this. I've too much to wrap up in the next five work days. It looks like there's a mill issue currently. Otherwise I think it should be ready to merge.

**Comment by danking** [comment]
OK, I'm not sure how to fix this but the work is to explain to the GCS Hadoop Connector which credentials we want it to use.  See the failure here: https://batch.hail.is/batches/8136069/jobs/49 . It uses CI's credentials instead of the test credentials. We use core-site.xml to do this in Spark <3.5, but the GCS connector is different in Spark 3.5 and it uses different configuration parameters. My most recent change did not successfully configure it.

Daniel G can help you a bit with credentials in Batch if that's necessary but the real work is to figure out how to tell the GCS Hadoop Connector to use the /gsa-key/key.json file.

**Comment by danking** [comment]
Just the local backend is broken now. I might've just fixed that.

**Comment by danking** [comment]
OK, I won't be able to fix this. @ehigham @patrick-schultz @daniel-goldstein some combo of you three can probably figure it out. The local backend tests that hit requester pays buckets are failing with new Spark. New Spark needs new GCS hadoop connector (see the Dockerfiles). New GCS hadoop connector has [brand new configuration parameters](https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/v3.0.0/gcs/INSTALL.md). Somehow I managed to make the normal Spark backend work correctly but the Local backend (which still, afaik, uses Spark & Hadoop for filesystems) is still trying to pick up CI's credentials instead of the test account's credentials.

```
E           hail.utils.java.FatalError: GoogleJsonResponseException: 403 Forbidden
E           GET https://storage.googleapis.com/storage/v1/b/hail-test-requester-pays-fds32/o/zero-to-nine?fields=bucket,name,timeCreated,updated,generation,metageneration,size,contentType,contentEncoding,md5Hash,crc32c,metadata&userProject=hail-vdc
E           {
E             "code": 403,
E             "errors": [
E               {
E                 "domain": "global",
E                 "message": "ci-910@hail-vdc.iam.gserviceaccount.com does not have serviceusage.services.use access to the Google Cloud project. Permission 'serviceusage.services.use' denied on resource (or it may not exist).",
E                 "reason": "forbidden"
E               }
E             ],
E             "message": "ci-910@hail-vdc.iam.gserviceaccount.com does not have serviceusage.services.use access to the Google Cloud project. Permission 'serviceusage.services.use' denied on resource (or it may not exist)."
E           }
E           
E           Java stack trace:
E           java.io.IOException: Error accessing gs://hail-test-requester-pays-fds32/zero-to-nine
E           	at com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.GoogleCloudStorageImpl.getObject(GoogleCloudStorageImpl.java:1986)
E           	at com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.GoogleCloudStorageImpl.getItemInfo(GoogleCloudStorageImpl.java:1882)
E           	at com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.GoogleCloudStorageFileSystemImpl.getFileInfoInternal(GoogleCloudStorageFileSystemImpl.java:861)
E           	at com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.GoogleCloudStorageFileSystemImpl.getFileInfo(GoogleCloudStorageFileSystemImpl.java:833)
E           	at com.google.cloud.hadoop.fs.gcs.GoogleHadoopFileSystem.getFileStatus(GoogleHadoopFileSystem.java:724)
E           	at org.apache.hadoop.fs.Globber.getFileStatus(Globber.java:115)
E           	at org.apache.hadoop.fs.Globber.doGlob(Globber.java:349)
E           	at org.apache.hadoop.fs.Globber.glob(Globber.java:202)
E           	at org.apache.hadoop.fs.FileSystem.globStatus(FileSystem.java:2142)
E           	at com.google.cloud.hadoop.fs.gcs.GoogleHadoopFileSystem.globStatus(GoogleHadoopFileSystem.java:759)
E           	at com.google.cloud.hadoop.fs.gcs.GoogleHadoopFileSystem.globStatus(GoogleHadoopFileSystem.java:1277)
E           	at is.hail.io.fs.HadoopFS.glob(HadoopFS.scala:162)
E           	at is.hail.io.fs.HadoopFS.glob(HadoopFS.scala:85)
E           	at is.hail.io.fs.FS.glob(FS.scala:402)
E           	at is.hail.io.fs.FS.glob$(FS.scala:402)
E           	at is.hail.io.fs.HadoopFS.glob(HadoopFS.scala:85)
E           	at is.hail.io.fs.HadoopFS.$anonfun$globAll$1(HadoopFS.scala:154)
E           	at is.hail.io.fs.HadoopFS.$anonfun$globAll$1$adapted(HadoopFS.scala:153)

```

File: docker/core-site.xml
Comment (daniel-goldstein) [inline docker/core-site.xml:15]: These properties look wrong based on [this example](https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/b0fff1c6e45e6c0bd74fc7ce228a9f86da439ce8/gcs/src/test/resources/core-site.xml#L58-L65)

Comment (patrick-schultz) [reply docker/core-site.xml:15]: Is this still wrong?

File: hail/Makefile
Comment (patrick-schultz) [inline hail/Makefile:534]: Can we leave a comment here explaining that this is deleting something that's only created by the old gradle build?

File: docker/core-site.xml
Comment (ehigham) [reply docker/core-site.xml:15]: Dan fixed this, hence spark tests pass

Comment (ehigham) [reply docker/core-site.xml:15]: I don't know what to say - seems to work in ci?

Comment (ehigham) [reply docker/core-site.xml:15]: This is what [`Install.md`](https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/af94c95d817f4dc0ce8c4094ce2445a8cc0a1d37/gcs/INSTALL.md?plain=1#L75-L82) recommends

Comment (daniel-goldstein) [reply docker/core-site.xml:15]: *shrug* guess it's fine then

**Review by patrick-schultz — CHANGES_REQUESTED** [review]
Can you just change the hail version in the changelog text before we merge it?

**Comment by ehigham** [comment]
> I'm good with this, just one minor request, and a resolution on @daniel-goldstein 's question above.
> 
> Also, need to update the hail version in the changelog text

Sorry, missed that bit about the changelog

**Review by patrick-schultz — APPROVED** [review]

**Pull request merged**
